### PR TITLE
Monitor service monitor more stats

### PR DESCRIPTION
**Pull request by tjoneslo**
This merge does two things. 

It add a collection of counter to the SessionMonitor for counting the types of statements being run through the Postgres Interface. The counters enhance the "Statements" count with Failed, From Cache, and Logged statments, plus a count of the different types of statements : Select, DML, DDL, call, copy, explain, emulated metadata, alter session, and alter server calls. These statistics are exposed through additional columns to the IS.SERVER_SESSIONS table.

Second it makes the MonitorService also collect the same set of statistics from all the sessions. These are exposed through a new IS.SERVER_STATISTICS_SUMMARY table.

### CODE REVIEW

File: fdb-sql-layer-pg/src/main/java/com/foundationdb/sql/pg/PostgresServerConnection.java [hunk 1]
```diff
@@ -695,9 +695,11 @@ protected void processQuery() throws IOException {
         updateAIS(context);
         
         PostgresStatement pstmt = null;
-        if (statementCache != null)
+        if (statementCache != null) 
             pstmt = statementCache.get(sql);
-        if (pstmt == null) {
+        if (pstmt != null) {
+            sessionMonitor.countEvent(StatementTypes.FROM_CACHE);
```
Comment (ScottDugas) [inline fdb-sql-layer-pg/src/main/java/com/foundationdb/sql/pg/PostgresServerConnection.java]: What about prepared versus unprepared versus preparing?

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/MonitorServiceImpl.java [hunk 2]
```diff
@@ -63,6 +63,8 @@
     
     private Map<String, UserMonitor> users;
 
+    private ConcurrentHashMap<StatementTypes, AtomicLong> statementCounter;
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/MonitorServiceImpl.java]: An AtomicLong[] indexed by StatementTypes.ordinal() probably makes more sense in this situation.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionEventListener.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Copyright (C) 2009-2013 FoundationDB, LLC
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionEventListener.java]: 2014

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitor.java [hunk 4]
```diff
@@ -79,4 +85,35 @@
...
+        SESSION    ("Session statement"),
+        SERVER     ("Server statement");
+        
+        private StatementTypes (String name) {
+            this.name = name;
+        }
+        private String name; 
+        
+        public String toString() {
+            return name;
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitor.java]: The names given here don't really provide anything the default toString() doesn't.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitorBase.java [hunk 5]
```diff
@@ -31,9 +33,12 @@
     private long currentStatementStartTime = -1;
     private long currentStatementEndTime = -1;
     private int rowsProcessed = 0;
-    private int statementCount = 0;
     private UserMonitor user = null; 
-
+    
+    private StatementCounter<StatementTypes, Long> statementCounters = new StatementCounter<StatementTypes, Long>();
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitorBase.java]: A long[] indexed by StatementTypes.ordinal() would probably be just fine here.

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitorBase.java [hunk 6]
```diff
@@ -31,9 +33,12 @@
...
+    private List<SessionEventListener> events = new CopyOnWriteArrayList<>();
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitorBase.java]: eventListeners would be more accurate. Is the copy-on-write needed?

**Comment by nathanlws** [comment]
A few simple code comments but looks mostly as expected.

I'm not sure we want to directly echo the names of the internal statement types out to the user. Perhaps just SELECT, DML, DDL and OTHER for statement type breakdown?

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitor.java [hunk 7]
```diff
@@ -79,4 +85,35 @@
     
     /** Get the user monitor for this session */
     UserMonitor getUserMonitor();
+
+    public void addSessionEventListener(SessionEventListener listener);
+    public void removeSessionEventListener (SessionEventListener listener);
+    
+    
+    public enum StatementTypes {
+        STATEMENT  ("Statement"),
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitor.java]: Would STARTED be clearer?

File: fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitor.java [hunk 8]
```diff
@@ -79,4 +85,35 @@
...
+        FAILED     ("Failed statement"),
+        FROM_CACHE ("Statement from cache"),
+        LOGGED     ("Statment logged"),
```
Comment (nathanlws) [inline fdb-sql-layer-core/src/main/java/com/foundationdb/server/service/monitor/SessionMonitor.java]: I'm not sure that LOGGED is all that useful unless we call it SLOW, though that will be overlapping with the new i_s coming soon.

File: fdb-sql-layer-pg/src/main/java/com/foundationdb/sql/pg/PostgresServerConnection.java (hunk 1 above)
Comment (tjoneslo) [inline fdb-sql-layer-pg/src/main/java/com/foundationdb/sql/pg/PostgresServerConnection.java]: I think for right now I have a good balance of statistics. Create a Trello card if you think we need more.

Comment (ScottDugas) [inline fdb-sql-layer-pg/src/main/java/com/foundationdb/sql/pg/PostgresServerConnection.java]: OK.

**Comment by nathanlws** [comment]
Thanks for the tweaks.

Approve

**Pull request merged**